### PR TITLE
Removed IToolResolver interface.

### LINKED ISSUE #337: Remove IToolResolver interface
It's over engineered and doesn't accomplish much. The only tool that uses this is the NuGet tool and this could easily be exposed from the ICakeContext since some core functionality depends on it.

### PR DESCRIPTION
**Pull request by patriksvensson**
Since it's only `INuGetToolResolver` using this, this reduces the complexity in general.
Also added shared base class for NuGet tools (`NuGetTool<TSettings>`) to reduce redundancy.

Closes #337

### CODE REVIEW

File: src/Cake/Cake.csproj [hunk 1]
```diff
@@ -42,6 +42,7 @@
     <RunCodeAnalysis>false</RunCodeAnalysis>
     <CodeAnalysisRuleSet>..\Cake.ruleset</CodeAnalysisRuleSet>
     <Prefer32Bit>false</Prefer32Bit>
+    <LangVersion>default</LangVersion>
```
Comment (devlead) [inline src/Cake/Cake.csproj]: As there's a PR for  ` <LangVersion>5</LangVersion>` this change feels unnecessary, right?

File: src/Cake.Tests/Cake.Tests.csproj [hunk 2]
```diff
@@ -23,6 +23,7 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
+    <LangVersion>default</LangVersion>
```
Comment (devlead) [inline src/Cake.Tests/Cake.Tests.csproj]: As there's a PR for  ` <LangVersion>5</LangVersion>` this change feels unnecessary, right?

File: src/Cake.Testing/Cake.Testing.csproj [hunk 3]
```diff
@@ -24,6 +24,7 @@
     <WarningLevel>4</WarningLevel>
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
     <CodeAnalysisRuleSet>..\Cake.ruleset</CodeAnalysisRuleSet>
+    <LangVersion>default</LangVersion>
```
Comment (devlead) [inline src/Cake.Testing/Cake.Testing.csproj]: As there's a PR for  ` <LangVersion>5</LangVersion>` this change feels unnecessary, right?

**Comment by devlead** [comment]
:+1: this really cleans up the code! Look fine by me.

File: src/Cake/Cake.csproj (hunk 1 above)
Comment (patriksvensson) [inline src/Cake/Cake.csproj]: Could have sworn I stashed this... Thanks :+1:

**Pull request merged**